docs/strings.md

```diff
@@ -0,0 +1,464 @@
+
+# Modification
+
+For mutable strings, inserting, deleting, and replacing substrings often need to be fast.
+This typically leads to mutable strings using a twine or similar representation.
+Even in-place modification of a character or a substring with a string of the same length can involve changing the number of code units in a string if the characters are replaced with different-width ones.
+This provides a strong incentive to avoid mandating contiguous storage for strings and is one of the common reasons for avoiding `std::string` in C++.
+
+# Comparison
+
+There are four common kinds of comparison on strings:
+
+ - Canonicalisation-unaware equality.
+   Do these two strings contain the same set of Unicode code points?
+ - Canonicalisation-aware equality.
+   Do these two strings contain a sequence of Unicode code points that are equal under canonicalisation.
+ - Locale-unaware ordered comparison.
+   What is a stable ordering between these two strings that is fast to compute?
+ - Locale-aware collation.
+   What is the ordering between these two strings, with the comparison computed using the Unicode collation rules?
+
+The first and third of these depend only on the strings being able to expose length and iterators that export a common character encoding.
+The second and fourth require a complex unicode library and should probably be in a separate package.
+
+# URL and path operations
+
+Some frameworks provide operations on strings for handling path and URL data.
+These are increasingly being deprecated and being moved to specific path and URL types that enforce validity and use structured storage until they must be exported to a string.
+It is unlikely that a Verona string design will want to include any of these operations in the core string APIs.
+
+Encodings
+---------
+
+Unicode defines three common serialisations, with different properties:
+
+ - UTF-8 is the most space-efficient for most Latin alphabet languages (an Emoji) and encodes every Unicode code point in 1-4 bytes.
+   The variation in length can make processing and random access indexed by code point expensive.
+ - UTF-16 is the most space-efficient for CJK languages and encodes every Unicode code point to 2 or 4 bytes.
+   As with UTF-8, random access requires additional metadata, though this can typically be omitted for strings in European languages, where every non-Emoji character will be 2 bytes.
+ - UTF-32 is the least space-efficient encoding, using 4 bytes for every character, but has the advantage of being a fixed-length encoding.
+
+UTF-8 is common on a lot of Internet protocols and so may be the input or output format.
+UTF-8 is typically used for system call or C library arguments on UNIX systems, whereas UTF-16 is common on Windows.
+
+7-bit ASCII is sufficient to encode anything that can be encoded as a fixed-width UTF-8 string and is commonly sufficient for path and URL encodings and so can be worth special-case handling.
```

Comment:
As I recall (and it's been a long time since I read the relevant RFCs), URLs are actually quite a small subset of ASCII.  We definitely want to be able to convert URLs to and from strings, but I think it's a mistake to use strings to encode URLs for processing.  You want structured data that contains the schema, password, domain, port, path, and so on, rather than a string that you do operations that match on character ranges on.